akka-projection-grpc/src/main/scala/akka/persistence/EventWriter.scala

```diff
@@ -0,0 +1,201 @@
+              perPidWriteState = perPidWriteState.updated(pid, newStateForPid)
+            } else {
+              if (newStateForPid.waitingForWrite.isEmpty) {
+                perPidWriteState = perPidWriteState - pid
+              } else {
+                // batch waiting for pid
+                if (context.log.isTraceEnabled())
+                  context.log.traceN(
+                    "Writing batch of {} events for pid [{}], seq nrs [{}-{}]",
+                    newStateForPid.waitingForWrite.size,
+                    pid,
+                    newStateForPid.waitingForWrite.head._1.sequenceNr,
+                    newStateForPid.waitingForWrite.last._1.sequenceNr)
+                val batchWrite = AtomicWrite(newStateForPid.waitingForWrite.map { case (repr, _) => repr })
+                journal ! JournalProtocol
+                  .WriteMessages(batchWrite :: Nil, context.self.toClassic, actorInstanceId)
+
+                val newReplyTo = newStateForPid.waitingForWrite.map {
+                  case (repr, replyTo) => repr.sequenceNr -> replyTo
+                }.toMap
+                perPidWriteState = perPidWriteState.updated(pid, StateForPid(newReplyTo, emptyWaitingForWrite))
+              }
+            }
+          }
+
+          def handleJournalResponse(response: JournalProtocol.Response): Behavior[AnyRef] =
+            response match {
+              case JournalProtocol.WriteMessageSuccess(message, `actorInstanceId`) =>
+                val pid = message.persistenceId
+                val sequenceNr = message.sequenceNr
+                perPidWriteState.get(pid) match {
+                  case None =>
+                    throw new IllegalStateException(
+                      s"Got write success reply for event with no waiting request, probably a bug (pid $pid, seq nr $sequenceNr)")
+                  case Some(stateForPid) =>
+                    stateForPid.waitingForReply.get(sequenceNr) match {
+                      case None =>
+                        throw new IllegalStateException(
+                          s"Got write success reply for event with no waiting request, probably a bug (pid $pid, seq nr $sequenceNr)")
+                      case Some(waiting) =>
+                        if (context.log.isTraceEnabled)
+                          context.log.trace2(
+                            "Successfully wrote event persistence id [{}], sequence nr [{}]",
+                            pid,
+                            message.sequenceNr)
+                        waiting ! StatusReply.success(WriteAck(pid, sequenceNr))
+                        val newState = stateForPid.copy(waitingForReply = stateForPid.waitingForReply - sequenceNr)
+                        handleUpdatedStateForPid(pid, newState)
+                        Behaviors.same
+                    }
+                }
+
+              case JournalProtocol.WriteMessageFailure(message, error, `actorInstanceId`) =>
```

Comment:
How should we handle duplicates, which will result in write failures? We will have duplicates since it's at-least-once.

One idea is to optimistically write, but if it fails we can check if the pid/seqNr exists with `EventTimestampQuery` or `LoadEventQuery`, or even `eventsByPersistenceId` since we are interested in more than one in the batch.

Not great to add a dependency to readJournalPluginId too, so maybe we can do this with the journal protocol?
asyncReadHighestSequenceNr is not exposed as journal protocol message, so if we don't want to change Akka we could use `ReplayMessages` to find the highest sequence number.

Reply:
Hmm, and the exception is from the plugin, so there is no way to determine that it is a duplicate-write from the failure response. Replay seems like a good trick, because we could do fromSequenceNr = toSequenceNr just for that seqNr.

Reply:
Exception is from plugin, but we can't see the difference of duplicate error to any other error, at this level.
Some plugins may not fail on duplicates, such as Cassandra, but that is another story.

and I think you know which other seqNrs are pending from the same batch, and it's very likely that they will also fail. Then it's probably easier to use WriteMessagesFailure to trigger these checks. We know that all or none will fail because they are in AtomicWrite.